### DISCUSSION TITLE
Setting/updating atmospheric properties

### DISCUSSION THREAD

**Discussion by benbusby** [JSBSim-Team/jsbsim#306] [Support requests]
Hi,

I've been working on developing a good way to dynamically set and update atmospheric properties through the Python API, but have been running into a couple of issues. 

It seems that at least some of the documented atmospheric properties in JSBSim (see [here](https://github.com/bstaero/jsbsim/blob/master/properties.md#atmosphere)) might be immutable, since attempts to set new a value for atmospheric properties don't seem to actually update the value. Here's an excerpt of how I'm attempting to set temperature, for example:

```python
self.flight_model.set_property_value('atmosphere/T-R', 540.27)
print(self.flight_model.get_property_value('atmosphere/T-R'))

# Prints '518.655...'
```

Are these atmospheric properties intended to accept user modification or are they meant to be immutable? 

In addition, I'd also like to be able to set more properties than what I've come across in the [sourceforge documentation](http://jsbsim.sourceforge.net/JSBSim/classJSBSim_1_1FGAtmosphere.html), namely things like gravity, wind speed, etc, and eventually integrate global reference atmospheric models like the Venus-GRAM into my use case, but the properties listed don't seem to include these. Is there a separate reference for atmospheric properties that I'm overlooking or is this the full list?

Thanks again for all of your work on this project!

**Comment by seanmcleod** [comment]
Take a look at this set of properties, which all have setters and getters.

https://github.com/JSBSim-Team/jsbsim/blob/76411ec4d0a095ca031da320c466db3561c31267/src/models/atmosphere/FGStandardAtmosphere.cpp#L671-L698

Then the following set only have getters.

https://github.com/JSBSim-Team/jsbsim/blob/76411ec4d0a095ca031da320c466db3561c31267/src/models/FGAtmosphere.cpp#L271-L286

**Comment by benbusby** [comment]
Thanks for the quick response! So despite `FGAtmosphere` properties not having setters, could I not just use the `atmosphere/override/<property name>` nodes to achieve the same effect? I'm referencing this section (although it still looks like a lot of the control I was hoping for is not available):

https://github.com/JSBSim-Team/jsbsim/blob/76411ec4d0a095ca031da320c466db3561c31267/src/models/FGAtmosphere.cpp#L107-L131

**Comment by seanmcleod** [comment]
I guess so, I didn't spot those. I've only used `atmosphere/delta-T` in the past.

If they still don't offer the level of control you require then I guess your other option is to implement your own atmosphere model. In addition to the standard atmosphere model you'll see there are 2 other atmosphere models currently in JSBSim, MSIS and Mars.

**Reply by beltzjr** [reply]
How can the MSIS or Mars models be chosen for a run? Haven't been able to find any example.

**Reply by bcoconni** [reply]
Unfortunately at the moment the Mars and MSIS atmosphere models are dead code, you cannot select them.

**Reply by beltzjr** [reply]
Can you give more background in this? Does someone need to just fix it? Is there a master plan in work for how to better implement it?

**Reply by bcoconni** [reply]
This feature could be added with a reasonable amount of effort (C++ coding involved). At the moment we are processing an element `<planet/>` which allows to alter the planet geometry and gravity:

https://github.com/JSBSim-Team/jsbsim/blob/a35be5ce6f7d7ca0ef9a11a62563e54d769bc4f4/src/FGFDMExec.cpp#L801-L812

In order to allow modifying the atmosphere itself, you'd need to process an subelement `<atmosphere/>` in `<planet/>` and modify the atmosphere accordingly. The class `FGAtmosphere` would need a new method `Load` that would select the atmosphere.

However rather than using the `FGMars` class which design is quite out fashioned with respect to `FGStandardAtmosphere`, I'd rather inherit from `FGStandardAtmosphere` and allow customization of the table `StdAtmosTemperatureTable` and the likes.

**Comment by bcoconni** [comment]
Properties `atmosphere/override/*` are meant to bypass the ISA1976 atmospheric model, rather than letting JSBSim compute the pressure and temperature from the altitude and the atmospheric properties at sea level. When using these properties, JSBSim assumes your program is taking over the atmospheric properties computations.

As @seanmcleod mentioned, the temperature can be modified by the following properties:
* property `atmosphere/delta-T`  which basically shifts the temperature by the amount specified (in Rankines)
* property `atmosphere/SL-graded-delta-T` which modifies the slope of the temperature as a function of altitude.

When fed with these 2 properties, JSBSim assumes the ISA1976 atmospheric model must be used to assess the pressure and temperature.

**Comment by bcoconni** [comment]
This issue has been moved to "Discussions". Now that this new feature is available in GitHub, it has been decided to move issues related to support requests and questions to "Discussions" and restrict "Issues" to bugs and feature requests.
This change is only motivated by project organization and will not affect the support provided. We'll gladly continue to answer your questions 😄.
